### PR TITLE
Add cilium network policies

### PR DESCRIPTION
**Pull request by khewonc**
### What does this PR do?

Add cilium network policy manager as a follow up to implementing `ApplyGlobalSettings()`

Note: does not include feature-specific policies

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

* Install Cilium in a cluster: [kind cluster example using cilium v1.12](https://docs.cilium.io/en/v1.12/gettingstarted/kind/)
* Spin up operator + DatadogAgent and check that cilium network policies are created
```yaml
spec:
  global:
    networkPolicy:
      create: true
      flavor: cilium
```
```
> kubectl get ciliumnetworkpolicy   
NAME                            AGE
datadog-agent                   14s
datadog-cluster-agent           14s
datadog-cluster-checks-runner   14s
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -125,4 +125,4 @@ require (
 	sigs.k8s.io/kustomize/api v0.10.1 // indirect
 	sigs.k8s.io/kustomize/kyaml v0.13.0 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
-)
+)
```
Comment (khewonc) [inline go.mod:128]: Will add back the newline

File: go.sum [hunk 2]
```diff
@@ -1198,4 +1198,4 @@ sigs.k8s.io/structured-merge-diff/v4 v4.2.1/go.mod h1:j/nl6xW8vLS49O8YvXW1ocPhZa
 sigs.k8s.io/yaml v1.1.0/go.mod h1:UJmg0vDUVViEyp3mgSv9WPwZCDxu4rQW1olrI1uml+o=
 sigs.k8s.io/yaml v1.2.0/go.mod h1:yfXDCHCao9+ENCvLSE62v9VSji2MKu5jeNfTrofGhJc=
 sigs.k8s.io/yaml v1.3.0 h1:a2VclLzOGrwOHDiV8EfBGhvjHvP46CtW5j6POvhYGGo=
-sigs.k8s.io/yaml v1.3.0/go.mod h1:GeOyir5tyXNByN85N/dRIT9es5UQNerPYEKK56eTBm8=
+sigs.k8s.io/yaml v1.3.0/go.mod h1:GeOyir5tyXNByN85N/dRIT9es5UQNerPYEKK56eTBm8=
```
Comment (khewonc) [inline go.sum:1201]: Same with the newline here

File: controllers/datadogagent/override/global.go [hunk 3]
```diff
@@ -7,15 +7,44 @@ package override
...
+	// NetworkPolicy contains the network configuration.
+	if config.NetworkPolicy != nil {
+		if apiutils.BoolValue(config.NetworkPolicy.Create) {
+			switch config.NetworkPolicy.Flavor {
+			case v2alpha1.NetworkPolicyFlavorCilium:
+				resourcesManager.CiliumPolicyManager().SetDDASite(*config.Site)
+				if config.Endpoint != nil && *config.Endpoint.URL != "" {
+					resourcesManager.CiliumPolicyManager().SetDDAURL(*config.Endpoint.URL)
+				}
+				if _, ok := dda.Spec.Override[v2alpha1.ClusterAgentComponentName]; ok {
```
Comment (celenechang) [inline controllers/datadogagent/override/global.go:31]: i guess we should also check if the component is not disabled

Comment (celenechang) [reply controllers/datadogagent/override/global.go:31]: actually, let me think about this a bit

File: controllers/datadogagent/merger/cilium.go [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+
+func (m *ciliumPolicyManagerImpl) SetupCiliumManager(site, ddURL string, hostNetwork bool, dnsSelectorEndpoints []metav1.LabelSelector) {
+	m.site = site
+	m.ddURL = ddURL
+	m.hostNetwork = hostNetwork
+	m.dnsSelectorEndpoints = dnsSelectorEndpoints
+}
+
+// AddCiliumPolicies creates or updates multiple cilium network policies
+func (m *ciliumPolicyManagerImpl) CreateCiliumPolicy(name, namespace string, policySpecs []cilium.NetworkPolicySpec) error {
```
Comment (celenechang) [inline controllers/datadogagent/merger/cilium.go:59]: nit the name in the comment doesn't match function name. (perhaps use AddCiliumPolicy to be consistent with the kubernetes network policy version?)

File: controllers/datadogagent/merger/cilium.go [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+			egressDNS(podSelector, m.dnsSelectorEndpoints),
+			egressCCRDatadogIntake(podSelector, m.site, m.ddURL),
+			egressCCRToDCA(podSelector, dda),
+			egressChecks(podSelector),
+		}
+	}
+	return m.CreateCiliumPolicy(policyName, dda.GetNamespace(), policySpecs)
+}
+
+func getPolicyMetadata(dda metav1.Object, componentName v2alpha1.ComponentName) (policyName string, podSelector metav1.LabelSelector) {
```
Comment (celenechang) [inline controllers/datadogagent/merger/cilium.go:124]: if i'm reading correctly, this is identical to the function of the same name in merger/network_policy.go. should we move to a common file?

File: controllers/datadogagent/component/utils.go [hunk 6]
```diff
@@ -478,3 +487,504 @@ func ShouldCreateAgentLocalService(gitVersion string, forceEnableLocalService bo
...
+			egressMetadataServerRule(podSelector),
+			egressDNS(podSelector, dnsSelectorEndpoints),
+			egressAgentDatadogIntake(podSelector, site, ddURL),
+			egressKubelet(podSelector),
+			ingressDogstatsd(podSelector),
+			egressChecks(podSelector),
+		}
+	case v2alpha1.ClusterAgentComponentName:
+		_, nodeAgentPodSelector := GetNetworkPolicyMetadata(dda, v2alpha1.NodeAgentComponentName)
+		policySpecs = []cilium.NetworkPolicySpec{
```
Comment (davidor) [inline controllers/datadogagent/component/utils.go:510]: Similar to the above, I think we're missing an ingress rule when cluster runners are enabled, and another one when external metrics are enabled.

File: controllers/datadogagent/component/utils.go [hunk 7]
```diff
@@ -478,3 +487,504 @@ func ShouldCreateAgentLocalService(gitVersion string, forceEnableLocalService bo
...
 }
+
+// BuildCiliumPolicy creates the base node agent, DCA, or CCR cilium network policy
+func BuildCiliumPolicy(dda metav1.Object, site string, ddURL string, hostNetwork bool, dnsSelectorEndpoints []metav1.LabelSelector, componentName v2alpha1.ComponentName) (string, string, []cilium.NetworkPolicySpec) {
+	policyName, podSelector := GetNetworkPolicyMetadata(dda, componentName)
+	var policySpecs []cilium.NetworkPolicySpec
+
+	switch componentName {
+	case v2alpha1.NodeAgentComponentName:
+		policySpecs = []cilium.NetworkPolicySpec{
```
[the remark this answers is not in the document]
Comment (khewonc) [reply controllers/datadogagent/component/utils.go:498]: The feature-specific cilium (and kubernetes) network policies will be handled separately in each feature's `feature.go`. I haven't gotten around to those yet but they're on my todo list. This PR only attempts to add policies that should be present by default

[the remark this answers is not in the document]
Comment (davidor) [reply controllers/datadogagent/component/utils.go:498]: Thanks for clarifying :+1:

**Review by davidor — APPROVED** [review]

**Pull request merged**